### PR TITLE
Enable checking for CLI maintenance

### PR DESCRIPTION
**Pull request by nhoobin**

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -46,8 +46,46 @@
...
+function checkclimaintenance($configfile) {
+    $content = file_get_contents($configfile);
+    $content = preg_replace("/\/\//", "\n//", $content);  // Set comments to be on newlines, replace '//' with '\n//'
+    $content = preg_replace("/;/", ";\n", $content);      // Split up statements, replace ';' with ';\n'
+    $content = preg_replace("/^[\s]+/m", "", $content);   // Removes all initial whitespace and newlines.
+
+    $lines = explode(PHP_EOL, $content);
+    foreach ($lines as $line) {
+        $line = trim($line);
+        $re = '/^\$CFG->dataroot\s+=\s+["\'](.*)["\'];/';  // Lines starting with $CFG->dataroot
```
Comment (brendanheywood) [inline index.php]: Can you please test this with this which I think will break it:

```$CFG->dataroot = "/bar"; # this used to be "/foo";```

Try changing ```(.*)``` to non-greedy ```(.*?)```

File: index.php [hunk 2]
```diff
@@ -46,8 +46,46 @@
...
+
+$climaintenaceon = checkclimaintenance($configfile);
+if ($climaintenaceon === false) {
+    require_once('../../../config.php');
+    global $CFG;
+} else {
+    $CFG = new stdClass();
+    $CFG->dataroot = $climaintenaceon;
+    // Cannot fullcheck with cli maintenance enabled.
+    $fullcheck = false;
```
Comment (brendanheywood) [inline index.php]: Would it be easier to read the flow if this just checked for climaint == true, and then spat out the message and exited early?

File: index.php [hunk 3]
```diff
@@ -46,6 +46,40 @@
...
+    $content = preg_replace("/\/\//", "\n//", $content);  // Set comments to be on newlines, replace '//' with '\n//'
+    $content = preg_replace("/;/", ";\n", $content);      // Split up statements, replace ';' with ';\n'
+    $content = preg_replace("/^[\s]+/m", "", $content);   // Removes all initial whitespace and newlines.
+
+    $lines = explode(PHP_EOL, $content);
+    foreach ($lines as $line) {
+        $line = trim($line);
+        $re = '/^\$CFG->dataroot\s+=\s+["\'](.*?)["\'];/';  // Lines starting with $CFG->dataroot
+        preg_match($re, $line, $matches);
+        if (!empty($matches) && count($matches) == 2) {
```
Comment (brendanheywood) [inline index.php]: can you make this

if (!empty($matches) ) {

File: index.php [hunk 4]
```diff
@@ -46,6 +46,40 @@
...
+            $climaintenance = $matches[1] . '/climaintenance.html';
```
Comment (brendanheywood) [inline index.php]: $matches[count(matches)-1] 

or maybe

$matches[-1]

**Comment by brendanheywood** [comment]
awesome :+1:

File: index.php [hunk 5]
```diff
@@ -46,6 +46,36 @@
 define('NO_UPGRADE_CHECK', true);
 define('ABORT_AFTER_CONFIG', true);
 
+/**
+ * Checks if the command line maintenance mode has been enabled. Skip the config bootstrapping.
+ *
+ * @param string $configfile The relative path for config.php
+ * @return bool True if climaintenance.html is found.
+ */
+function checkclimaintenance($configfile) {
```
Comment (roperto) [inline index.php]: Moodle standards allow underscore in function name, so you could use `check_climaintenance()` if you wanted.

File: index.php [hunk 6]
```diff
@@ -46,6 +46,36 @@
...
+    $content = file_get_contents($configfile);
+    $content = preg_replace("/[^!:]\/\//", "\n//", $content);  // Set comments to be on newlines, replace '//' with '\n//', where // does not start with :
```
Comment (roperto) [inline index.php]: Using another regex delimiter could make it more readable:
`preg_replace("#[^!:]//#", "\n//", $content);`

File: index.php [hunk 7]
```diff
@@ -46,6 +46,36 @@
...
+
+        if (file_exists($climaintenance)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+if (checkclimaintenance('../../../config.php') === true) {
```
Comment (roperto) [inline index.php]: We can make it absolute: `check_climaintenance(__DIR__.'/../../../config.php')`

File: index.php [hunk 8]
```diff
@@ -46,6 +46,36 @@
...
+    print "Server is in MAINTENANCE<br>\n";
+    exit;
+}
+
 require_once('../../../config.php');
```
Comment (roperto) [inline index.php]: Absolute: `require_once(__DIR__.'/../../../config.php');`

**Pull request merged**